File: Challenge80/solution.py

def is_point_in_path(x: int, y: int, poly) -> bool:

    num = len(poly)
    j = num - 1
    c = False
    for i in range(num):
        if (x == poly[i][0]) and (y == poly[i][1]):
            # point is a corner
            return True
        if ((poly[i][1] > y) != (poly[j][1] > y)):
            slope = (x-poly[i][0])*(poly[j][1]-poly[i][1]) - (poly[j][0]-poly[i][0])*(y-poly[i][1])
            if slope == 0:
                return True
            if (slope < 0) != (poly[j][1] < poly[i][1]):
                c = not c
        j = i
    return c


for _ in range(int(input())):
    n = int(input())
    string = input()
    string2 = input()
    string = string.split()
    string2 = string2.split()
    sides = []
    sides2 = []
    for i, j in zip(string, string2):
        x, y = map(int, i.split(","))
        x1, y1 = map(int, j.split(","))
        sides.append((x, y))
        sides2.append((x1, y1))

    first_polygon_sides = []
    for i in range(len(sides)-1):
        side = sides[i]+sides[i+1]
        first_polygon_sides.append(side)
    first_polygon_sides.append(sides[-1]+sides[0])

    second_polygon_sides = []
    for i in range(len(sides2)-1):
        side = sides2[i]+sides2[i+1]
        second_polygon_sides.append(side)
    second_polygon_sides.append(sides2[-1]+sides2[0])

    ans = "false"
    br = False
    for side1 in first_polygon_sides:
        if br:
            break
        for side2 in second_polygon_sides:

            a,b,c,d = side1
            w, x, y, z = side2

            dx0 = c-a
            dx1 = y - w
            dy0 = d - b
            dy1 = z - x
            p0 = dy1 * (y-a)-dx1*(z-b)
            p1 = dy1 * (y-c)-dx1*(z-d)
            p2 = dy0 * (c-w)-dx0*(d-x)
            p3 = dy0 * (c-y)-dx0*(d-z)
        
            if (p0*p1 < 0) & (p2*p3 < 0):
                ans = "true"
                br = True
                break
        
    if ans == "true":
        print(ans)
    else:
        for i in sides2:
            if is_point_in_path(i[0], i[1], sides):
                ans = "true"
                break
        if ans == "true":
            print(ans)
        else:
            for side in sides:
                if is_point_in_path(side[0], side[1],sides2):
                    ans = "true"
                    break
            print(ans)

File: Challenge80/solution.cpp
#include <bits/stdc++.h>
using namespace std;
#define side_type tuple<int, int, int, int>
#define polygon_type vector<pair<int, int>>
#define pb push_back

void getInput(polygon_type &vec, int &N)
{
    string comma = ",";
    for (int j = 0; j < N; ++j)
    {
        string tmp;
        cin >> tmp;

        int pos;
        while ((pos = tmp.find(comma)) != string::npos)
        {
            vec.pb({stoi(tmp.substr(0, pos)), stoi(tmp.substr(pos + 1, tmp.size()))});
            tmp.clear();
        }
    }
}

void AddSides(polygon_type &polygon, vector<side_type> &sides)
{
    for (int i = 0; i < polygon.size() - 1; i++)
    {
        sides.pb(side_type(polygon[i].first, polygon[i].second, polygon[i + 1].first, polygon[i + 1].second));
    }
    sides.pb(side_type(polygon.back().first, polygon.back().second, polygon[0].first, polygon[0].second));
}

//ngl, this function is a c++ version of the python code from wiki
bool PIP(int x, int y, polygon_type &polygon)
{
    int j = polygon.size() - 1;
    bool c = false;
    for (int i = 0; i < polygon.size(); ++i)
    {
        if (x == polygon[i].first && y == polygon[i].second)
        {
            return true;
        }
        if (polygon[i].second > y != polygon[j].second > y)
        {
            int slope = (x - polygon[i].first) * (polygon[j].second - polygon[i].second) - (polygon[j].first - polygon[i].first) * (y - polygon[i].second);
            if (slope == 0)
            {
                return true;
            }
            if (slope < 0 != (polygon[j].second < polygon[i].second))
            {
                c = !c;
            }
        }
        j = i;
    }
    return c;
}

void solve()
{
    int N;
    cin >> N;
    polygon_type polygon1, polygon2;

    getInput(polygon1, N);
    getInput(polygon2, N);

    vector<side_type> first_polygon_sides, second_polygon_sides;

    AddSides(polygon1, first_polygon_sides);
    AddSides(polygon2, second_polygon_sides);

    bool br = false;
    double a, b, c, d, w, x, y, z, dx0, dx1, dy0, dy1;
    for (auto side1 : first_polygon_sides)
    {
        if (br)
        {
            break;
        }
        for (auto side2 : second_polygon_sides)
        {
            tie(a, b, c, d) = side1;
            tie(w, x, y, z) = side2;
            dx0 = c - a;
            dx1 = y - w;
            dy0 = d - b;
            dy1 = z - x;
            //this formula is from stackoverflow(only this formula)
            if ((dy1 * (y - a) - dx1 * (z - b)) * (dy1 * (y - c) - dx1 * (z - d)) < 0 && (dy0 * (c - w) - dx0 * (d - x)) * (dy0 * (c - y) - dx0 * (d - z)) < 0)
            {
                br = true;
                break;
            }
        }
    }
    if (br)
    {
        cout << "true" << endl;
    }
    else
    {
        for (auto side : polygon1)
        {
            if (PIP(side.first, side.second, polygon2))
            {
                br = true;
                break;
            }
        }
        if (br)
        {
            cout << "true" << endl;
        }
        else
        {
            for (auto side : polygon2)
            {
                if (PIP(side.first, side.second, polygon1))
                {
                    br = true;
                    break;
                }
            }
            string a = br == 0 ? "false" : "true";
            cout << a << endl;
        }
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    int T;
    cin >> T;
    for (int i = 0; i < T; ++i)
    {
        solve();
    }
}
